experimental/dds/tree2/src/core/tree/cursor.ts

```diff
@@ -363,6 +363,36 @@ export function forEachNode<TCursor extends ITreeCursor = ITreeCursor>(
 	}
 }
 
+/**
+ * Moves cursor to each child of the current field.  The optional arguments `start` and `end`
+ * can be used to restrict the range of children visited.
+ *
+ * `callbackFn` is invoked after moving the cursor to each child.  The callback must return
+ * `true` to continue the enumeration, or `false` to early exit.
+ *
+ * The `callbackFn` may move move the provided cursor, but must restore the cursor to the current
+ * child node in the enumeration before returning.
+ */
+export function forEachNodeIn<TCursor extends ITreeCursor = ITreeCursor>(
+	cursor: TCursor,
+	callbackFn: (cursor: TCursor) => boolean,
+	start = 0,
+	end = cursor.getFieldLength(),
+): void {
+	assert(cursor.mode === CursorLocationType.Fields, 0x3bd /* should be in fields */);
+
+	// The placement of `remaining > 0` in the expressions below avoids unnecessary
+	// attempts to move the cursor when 0 items remain in the enumeration.
+	for (
+		let remaining = end - start,
+			shouldContinue = remaining > 0 && cursor.firstNode() && cursor.seekNodes(start);
```

Comment:
I think this code combines hitting the end of the field with hitting the end of the requested number of nodes.
I'm not sure thats a good idea since its confusing (should hitting the end of the field when you requested more nodes than that be an error?) and also, those cases leave the cursor in different places so they should be handles separately.

Getting a count of a nodes to visit, returning if its zero, then calling cursor.enterNode and doing something like `while count > 0 { count--; visiteNode; assert(cursor.nextNode(), "expected node");} cursor.up();
 would be much simpler, always termination iteration the same way (less cases to test/implement) and always put the cursor back at the start.

Reply:
I agree this is excessively terse.